lib/X25519KeyPair.js

```diff
@@ -0,0 +1,321 @@
+      decode: base58.decode,
+      keyMaterial: publicKeyBase58,
+      type: 'public'
+    });
+    const buffer = new Uint8Array(1 + pubkeyBytes.length);
+    buffer[0] = 0x25;
+    buffer.set(pubkeyBytes, 1);
+    // prefix with `z` to indicate multi-base base58btc encoding
+    return `z${base58.encode(buffer)}`;
+  }
+
+  /**
+   * Creates an instance of X25519KeyPair from a key fingerprint.
+   *
+   * @param {string} fingerprint
+   * @returns {X25519KeyPair}
+   * @throws Unsupported Fingerprint Type.
+   */
+  static fromFingerprint({fingerprint}) {
+    // skip leading `z` that indicates base58 encoding
+    const buffer = base58.decode(fingerprint.substr(1));
+
+    // buffer is: 0x25 <public key bytes>
+    if(buffer[0] !== 0x25) {
+      throw new Error(`Unsupported Fingerprint Type: ${fingerprint}`);
+    }
+
+    return new X25519KeyPair({
+      publicKeyBase58: base58.encode(buffer.slice(1))
+    });
+  }
+
+  /**
+   * Derives a shared secret via a given public key, typically for use
+   * as one parameter for computing a shared key. It should not be used as
+   * a shared key itself, but rather input into a key derivation function (KDF)
+   * to produce a shared key.
+   * @param {LDKeyPair} publicKey - Remote key pair
+   * @throws {TypeError} On invalid base58 encoding of public or private keys
+   * @returns {String}
+   */
+  deriveSecret({publicKey}) {
+    const remotePubkeyBytes = base58Decode({
+      decode: base58.decode,
+      keyMaterial: publicKey.publicKeyBase58,
+      type: 'public'
+    });
+
+    const privateKeyBytes = base58Decode({
+      decode: base58.decode,
+      keyMaterial: this.privateKeyBase58,
+      type: 'private'
+    });
+
+    return nacl.scalarMult(privateKeyBytes, remotePubkeyBytes);
+  }
+
+  /**
+   * Generates and returns a multiformats encoded
+   * X25519 public key fingerprint (for use with cryptonyms, for example).
+   * @see https://github.com/multiformats/multicodec
+   *
+   * @returns {string} The fingerprint.
+   */
+  fingerprint() {
+    const {publicKeyBase58} = this;
+    return X25519KeyPair.fingerprintFromPublicKey({publicKeyBase58});
+  }
+
+  /**
+   * Tests whether the fingerprint was generated from a given key pair.
+   * @example
+   * > xKeyPair.verifyFingerprint('...');
+   * {valid: true};
+   * @param {string} fingerprint - A Base58 public key.
+   *
+   * @returns {Object} An object indicating valid is true or false.
+   */
+  verifyFingerprint(fingerprint) {
+    // fingerprint should have `z` prefix indicating
+    // that it's multi-base encoded
+    if(!(typeof fingerprint === 'string' && fingerprint[0] === 'z')) {
+      return {
+        error: new Error('`fingerprint` must be a multibase encoded string.'),
+        valid: false
+      };
+    }
+    let fingerprintBuffer;
+    try {
+      fingerprintBuffer = base58Decode({
+        decode: base58.decode,
+        keyMaterial: fingerprint.slice(1),
+        type: `fingerprint's`
```

Comment:
I'll just note that's an odd looking "type".

Reply:
Agreed. I wasn't sure what was going on there.

Reply:
Oh... I see that's just for debugging.